// @build-tracker/comparator/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright (c) 2019 Paul Armstrong
 */
var build_1 = __importDefault(require("@build-tracker/build"));
var BuildDelta_1 = __importDefault(require("./BuildDelta"));
var markdown_table_1 = __importDefault(require("markdown-table"));
var types_1 = require("@build-tracker/types");
var formatting_1 = require("@build-tracker/formatting");
var CellType;
(function (CellType) {
    CellType["TEXT"] = "text";
    CellType["DELTA"] = "delta";
    CellType["TOTAL"] = "total";
    CellType["TOTAL_DELTA"] = "totalDelta";
    CellType["REVISION"] = "revision";
    CellType["REVISION_DELTA"] = "revisionDelta";
    CellType["ARTIFACT"] = "artifact";
    CellType["GROUP"] = "group";
})(CellType = exports.CellType || (exports.CellType = {}));
var emptyObject = Object.freeze({});
/* eslint-disable @typescript-eslint/no-explicit-any */
var flatten = function (arrays) { return arrays.reduce(function (memo, b) { return memo.concat(b); }, []); };
/* eslint-enable @typescript-eslint/no-explicit-any */
var defaultFormatCellText = function (cell) { return cell.text; };
var formatCellTextMarkdown = function (cell) { return cell.text.replace(/~/g, '\\~'); };
var defaultFormatRevision = function (cell) { return formatting_1.formatSha(cell.revision); };
var defaultFormatRevisionDelta = function (cell) { return "\u0394" + cell.deltaIndex; };
var defaultFormatTotal = function (cell, sizeKey) { return formatting_1.formatBytes(cell.sizes[sizeKey] || 0); };
var defaultFormatDelta = function (cell, sizeKey) {
    var errorFailingBudgets = cell.failingBudgets.some(function (result) { return result.level === types_1.BudgetLevel.ERROR; });
    var warningFailingBudgets = cell.failingBudgets.some(function (result) { return result.level === types_1.BudgetLevel.WARN; });
    return "" + (errorFailingBudgets
        ? formatting_1.levelToEmoji[types_1.BudgetLevel.ERROR] + " "
        : warningFailingBudgets
            ? formatting_1.levelToEmoji[types_1.BudgetLevel.WARN] + " "
            : cell.hashChangeUnexpected
                ? formatting_1.levelToEmoji['hash'] + " "
                : '') + formatting_1.formatBytes(cell.sizes[sizeKey] || 0) + " (" + ((cell.percents[sizeKey] || 0) * 100).toFixed(1) + "%)";
};
var defaultArtifactFilter = function () { return true; };
var deserializeRegExp = function (input) {
    var m = input.match(/\/(.*)\/(.*)?/);
    return new RegExp(m[1], m[2] || '');
};
var getFailingBudgetsFromRow = function (rows, level) {
    return rows.reduce(function (failures, row) {
        row.forEach(function (cell) {
            if (cell.type !== CellType.TOTAL_DELTA && cell.type !== CellType.DELTA) {
                return;
            }
            cell.failingBudgets.forEach(function (budget) {
                if (level && budget.level !== level) {
                    return;
                }
                failures.push({ name: row[0].text, result: budget });
            });
        });
        return failures;
    }, []);
};
var BuildComparator = /** @class */ (function () {
    function BuildComparator(_a) {
        var artifactBudgets = _a.artifactBudgets, artifactFilters = _a.artifactFilters, budgets = _a.budgets, builds = _a.builds, groups = _a.groups;
        this.builds = builds
            .sort(function (buildA, buildB) {
            var diff = buildA.timestamp.valueOf() - buildB.timestamp.valueOf();
            if (diff > 0) {
                return 1;
            }
            else if (diff < 0) {
                return -1;
            }
            return 0;
        })
            .sort(function (buildA, buildB) {
            if (buildA.getMetaValue('parentRevision') === buildB.getMetaValue('revision')) {
                return 1;
            }
            if (buildB.getMetaValue('revision') === buildA.getMetaValue('parentRevision')) {
                return -1;
            }
            return 0;
        });
        this._artifactFilters = artifactFilters || [];
        this._artifactBudgets = artifactBudgets || emptyObject;
        this._groups = __spreadArrays([{ name: 'All', artifactNames: this.artifactNames, budgets: budgets }], (groups || [])).filter(Boolean);
        this._emptySizes = Object.freeze(this.sizeKeys.reduce(function (memo, key) {
            memo[key] = 0;
            return memo;
        }, {}));
    }
    BuildComparator.deserialize = function (data) {
        var objData = JSON.parse(data);
        return new BuildComparator(__assign(__assign({}, objData), { artifactFilters: objData.artifactFilters.map(deserializeRegExp), builds: objData.builds.map(function (d) { return build_1.default.fromJSON(d); }), groups: objData.groups.map(function (g) { return (__assign(__assign({}, g), { artifactMatch: g.artifactMatch && deserializeRegExp(g.artifactMatch) })); }) }));
    };
    BuildComparator.prototype.serialize = function () {
        return JSON.stringify({
            artifactBudgets: this._artifactBudgets,
            artifactFilters: this._artifactFilters.map(function (f) { return f.toString(); }),
            budgets: this._groups[0].budgets,
            builds: this.builds.map(function (b) { return b.toJSON(); }),
            groups: this._groups.slice(1).map(function (g) { return (__assign(__assign({}, g), { artifactMatch: g.artifactMatch && g.artifactMatch.toString() })); }),
        });
    };
    Object.defineProperty(BuildComparator.prototype, "artifactNames", {
        get: function () {
            var _this = this;
            if (this.builds.length === 0) {
                return [];
            }
            if (!this._artifactNames) {
                this._artifactNames = Array.prototype.concat
                    .apply([], this.builds.map(function (build) { return build.artifacts; }))
                    .sort(function (a, b) {
                    var sizeKey = Object.keys(a.sizes)[0];
                    return b.sizes[sizeKey] - a.sizes[sizeKey];
                })
                    .map(function (artifact) { return artifact.name; })
                    .filter(function (value, index, self) {
                    return self.indexOf(value) === index && !_this._artifactFilters.some(function (filter) { return filter.test(value); });
                });
            }
            return this._artifactNames;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "sizeKeys", {
        get: function () {
            if (this.builds.length === 0 || this.artifactNames.length === 0) {
                return [];
            }
            if (!this._sizeKeys) {
                this._sizeKeys = Object.keys(this.builds[0].artifacts[0].sizes).sort();
                var allSizeKeys_1 = new Set();
                this.builds.forEach(function (build) {
                    build.artifacts.forEach(function (artifact) {
                        Object.keys(artifact.sizes).forEach(function (key) {
                            allSizeKeys_1.add(key);
                        });
                    });
                });
                if (allSizeKeys_1.size !== this._sizeKeys.length || !this._sizeKeys.every(function (key) { return allSizeKeys_1.has(key); })) {
                    throw new Error('builds provided do not have same size keys for artifacts');
                }
            }
            return this._sizeKeys;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "buildDeltas", {
        get: function () {
            var _this = this;
            if (this.builds.length === 0) {
                return [];
            }
            if (!this._buildDeltas) {
                this._buildDeltas = this.builds.map(function (baseBuild, i) {
                    return _this.builds.slice(0, i).map(function (prevBuild) {
                        return new BuildDelta_1.default(baseBuild, prevBuild, {
                            artifactBudgets: _this._artifactBudgets,
                            artifactFilters: _this._artifactFilters,
                            groups: _this._groups,
                        });
                    });
                });
            }
            return this._buildDeltas;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "matrixHeader", {
        get: function () {
            var _this = this;
            if (!this._matrixHeader) {
                this._matrixHeader = __spreadArrays([
                    { type: CellType.TEXT, text: '' }
                ], flatten(this.buildDeltas.map(function (buildDeltas, buildIndex) {
                    var revision = _this.builds[buildIndex].getMetaValue('revision');
                    return __spreadArrays([
                        { type: CellType.REVISION, revision: revision }
                    ], buildDeltas.map(function (buildDelta, deltaIndex) { return ({
                        type: CellType.REVISION_DELTA,
                        deltaIndex: deltaIndex + 1,
                        againstRevision: buildDelta.prevBuild.getMetaValue('revision'),
                        revision: revision,
                    }); }));
                })));
            }
            return this._matrixHeader;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "matrixGroups", {
        get: function () {
            var _this = this;
            if (!this._matrixGroups) {
                this._matrixGroups = this._groups.map(function (group) { return _this._getGroupRow(group); });
            }
            return this._matrixGroups;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "matrixArtifacts", {
        get: function () {
            var _this = this;
            if (!this._matrixArtifacts) {
                this._matrixArtifacts = this.artifactNames.map(function (artifactName) { return _this._getArtifactRow(artifactName); });
            }
            return this._matrixArtifacts;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "errors", {
        get: function () {
            if (!this._errors) {
                var groupResults = getFailingBudgetsFromRow(this.matrixGroups, types_1.BudgetLevel.ERROR);
                var artifactResults = getFailingBudgetsFromRow(this.matrixArtifacts, types_1.BudgetLevel.ERROR);
                this._errors = __spreadArrays(groupResults, artifactResults).filter(Boolean);
            }
            return this._errors;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "warnings", {
        get: function () {
            if (!this._warnings) {
                var groupResults = getFailingBudgetsFromRow(this.matrixGroups, types_1.BudgetLevel.WARN);
                var artifactResults = getFailingBudgetsFromRow(this.matrixArtifacts, types_1.BudgetLevel.WARN);
                this._warnings = __spreadArrays(groupResults, artifactResults).filter(Boolean);
            }
            return this._warnings;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BuildComparator.prototype, "unexpectedHashChanges", {
        get: function () {
            if (!this._unexpectedHashChanges) {
                this._unexpectedHashChanges = this.matrixArtifacts.reduce(function (memo, row) {
                    row.forEach(function (cell) {
                        if (cell.type !== CellType.DELTA) {
                            return;
                        }
                        if (cell.hashChangeUnexpected) {
                            memo.push({ name: cell.name });
                        }
                    });
                    return memo;
                }, []);
            }
            return this._unexpectedHashChanges;
        },
        enumerable: true,
        configurable: true
    });
    BuildComparator.prototype._getArtifactRow = function (artifactName) {
        var _this = this;
        var cells = this.buildDeltas.map(function (buildDeltas, i) {
            var artifact = _this.builds[i].getArtifact(artifactName);
            return __spreadArrays([
                {
                    sizes: artifact ? artifact.sizes : _this._emptySizes,
                    name: artifactName,
                    type: CellType.TOTAL,
                }
            ], buildDeltas.map(function (buildDelta) { return (__assign(__assign({}, buildDelta.getArtifactDelta(artifactName).toObject()), { type: CellType.DELTA })); }));
        });
        return __spreadArrays([{ type: CellType.ARTIFACT, text: artifactName }], flatten(cells));
    };
    BuildComparator.prototype._getGroupRow = function (group) {
        var _this = this;
        var artifactNames = group.artifactNames ? __spreadArrays(group.artifactNames).filter(Boolean) : [];
        if (group.artifactMatch) {
            artifactNames = artifactNames.concat(this.artifactNames.filter(function (name) { return group.artifactMatch.test(name); }));
        }
        var cells = this.buildDeltas.map(function (buildDeltas, i) {
            var groupSizes = _this.builds[i].getSum(artifactNames);
            return __spreadArrays([
                {
                    sizes: groupSizes,
                    name: group.name,
                    type: CellType.TOTAL,
                }
            ], buildDeltas.map(function (buildDelta) { return (__assign(__assign({}, buildDelta.getGroupDelta(group.name).toObject()), { type: CellType.TOTAL_DELTA })); }));
        });
        return __spreadArrays([{ type: CellType.GROUP, text: group.name, artifactNames: artifactNames }], flatten(cells));
    };
    BuildComparator.prototype.getStringFormattedHeader = function (formatRevision, formatRevisionDelta) {
        if (formatRevision === void 0) { formatRevision = defaultFormatRevision; }
        if (formatRevisionDelta === void 0) { formatRevisionDelta = defaultFormatRevisionDelta; }
        return this.matrixHeader.map(function (cell) {
            switch (cell.type) {
                case CellType.REVISION:
                    return formatRevision(cell);
                case CellType.REVISION_DELTA:
                    return formatRevisionDelta(cell);
                case CellType.TEXT:
                    return cell.text;
            }
        });
    };
    BuildComparator.prototype.getStringFormattedGroups = function (formatCellText, formatTotal, formatDelta, sizeKey) {
        if (formatCellText === void 0) { formatCellText = defaultFormatCellText; }
        if (formatTotal === void 0) { formatTotal = defaultFormatTotal; }
        if (formatDelta === void 0) { formatDelta = defaultFormatDelta; }
        if (sizeKey === void 0) { sizeKey = 'gzip'; }
        return this.matrixGroups.map(function (row) {
            return row.map(function (cell) {
                switch (cell.type) {
                    case CellType.GROUP:
                        return formatCellText(cell);
                    case CellType.TOTAL_DELTA:
                        return formatDelta(cell, sizeKey);
                    case CellType.TOTAL:
                        return formatTotal(cell, sizeKey);
                }
            });
        });
    };
    BuildComparator.prototype.getStringFormattedRows = function (formatCellText, formatTotal, formatDelta, sizeKey, artifactFilter) {
        if (formatCellText === void 0) { formatCellText = defaultFormatCellText; }
        if (formatTotal === void 0) { formatTotal = defaultFormatTotal; }
        if (formatDelta === void 0) { formatDelta = defaultFormatDelta; }
        if (sizeKey === void 0) { sizeKey = 'gzip'; }
        if (artifactFilter === void 0) { artifactFilter = defaultArtifactFilter; }
        return this.matrixArtifacts.filter(artifactFilter).map(function (row) {
            return row.map(function (cell) {
                switch (cell.type) {
                    case CellType.ARTIFACT:
                        return formatCellText(cell);
                    case CellType.DELTA:
                        return formatDelta(cell, sizeKey);
                    case CellType.TOTAL:
                        return formatTotal(cell, sizeKey);
                }
            });
        });
    };
    BuildComparator.prototype.toJSON = function () {
        return {
            header: this.matrixHeader,
            groups: this.matrixGroups,
            artifacts: this.matrixArtifacts,
        };
    };
    BuildComparator.prototype.toMarkdown = function (_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.formatCellText, formatCellText = _c === void 0 ? formatCellTextMarkdown : _c, formatRevision = _b.formatRevision, formatRevisionDelta = _b.formatRevisionDelta, formatTotal = _b.formatTotal, formatDelta = _b.formatDelta, artifactFilter = _b.artifactFilter, sizeKey = _b.sizeKey;
        var header = this.getStringFormattedHeader(formatRevision, formatRevisionDelta);
        var groups = this.getStringFormattedGroups(formatCellText, formatTotal, formatDelta, sizeKey);
        var rows = this.getStringFormattedRows(formatCellText, formatTotal, formatDelta, sizeKey, artifactFilter);
        var allRows = __spreadArrays([header], groups, rows);
        return markdown_table_1.default(allRows, { align: allRows[0].map(function (_, i) { return (i === 0 ? 'l' : 'r'); }) });
    };
    BuildComparator.prototype.toCsv = function (_a) {
        var _b = _a === void 0 ? {} : _a, formatCellText = _b.formatCellText, formatRevision = _b.formatRevision, formatRevisionDelta = _b.formatRevisionDelta, formatTotal = _b.formatTotal, formatDelta = _b.formatDelta, artifactFilter = _b.artifactFilter, sizeKey = _b.sizeKey;
        var header = this.getStringFormattedHeader(formatRevision, formatRevisionDelta);
        var groups = this.getStringFormattedGroups(formatCellText, formatTotal, formatDelta, sizeKey);
        var rows = this.getStringFormattedRows(formatCellText, formatTotal, formatDelta, sizeKey, artifactFilter);
        return __spreadArrays([header], groups, rows).map(function (row) { return "" + row.join(','); }).join("\r\n");
    };
    BuildComparator.prototype.toSummary = function (useEmoji) {
        if (useEmoji === void 0) { useEmoji = true; }
        var groupResults = getFailingBudgetsFromRow(this.matrixGroups).map(function (_a) {
            var name = _a.name, result = _a.result;
            return formatting_1.formatBudgetResult(result, "Group \"" + name + "\"", useEmoji);
        });
        var artifactResults = getFailingBudgetsFromRow(this.matrixArtifacts).map(function (_a) {
            var name = _a.name, result = _a.result;
            return formatting_1.formatBudgetResult(result, name, useEmoji);
        });
        var unexpectedHashChanges = this.unexpectedHashChanges.map(function (_a) {
            var name = _a.name;
            return formatting_1.formatUnexpectedHashChange(name, useEmoji);
        });
        var output = __spreadArrays(groupResults, artifactResults, unexpectedHashChanges).filter(Boolean);
        if (output.length === 0) {
            return [(useEmoji ? '✅' : 'Success:') + " No failing budgets"];
        }
        return output;
    };
    return BuildComparator;
}());
exports.default = BuildComparator;
